website/blog/2024/12-13-scrape-google-maps-using-python/index.md

```diff
@@ -0,0 +1,481 @@
+    print(f"Found {len(listings)} hotels\n")
+
+    # Process each hotel listing
+    for listing in listings:
+        # Extract details for each listing
+        data = {
+            "name": (
+                await (await listing.query_selector(".qBF1Pd")).inner_text()
+                if await listing.query_selector(".qBF1Pd")
+                else None
+            ),
+            "rating": (
+                await (await listing.query_selector(".MW4etd")).inner_text()
+                if await listing.query_selector(".MW4etd")
+                else "N/A"
+            ),
+            "reviews": (
+                (await (await listing.query_selector(".UY7F9")).inner_text())
+                .replace("(", "")
+                .replace(")", "")
+                if await listing.query_selector(".UY7F9")
+                else "N/A"
+            ),
+            "price": (
+                await (await listing.query_selector(".wcldff")).inner_text()
+                if await listing.query_selector(".wcldff")
+                else "N/A"
+            ),
+            "link": (
+                await (await listing.query_selector("a.hfpxzc")).get_attribute("href")
+                if await listing.query_selector("a.hfpxzc")
+                else "N/A"
+            ),
+            "amenities": [
+                await amenity.get_attribute("aria-label")
+                for amenity in await listing.query_selector_all(".dc6iWb")
+                if await amenity.get_attribute("aria-label")
+            ],
+        }
+
+        # Pretty-print the data
+        print(json.dumps(data, indent=4))
+        print("\n")
+```
+
+In the code:
+
+- `query_selector`: Returns first DOM element matching CSS selector, useful for single items like a name or rating
+- `query_selector_all`: Returns all matching elements, ideal for multiple items like amenities
+- `inner_text()`: Extracts text content
+- Some hotels might not have all the information available - we handle this with 'N/A’
+
+When you run this script, you'll see output similar to this:
+
+```json
+{
+    "name": "GRAND KALINGA HOTEL",
+    "rating": "4.2",
+    "reviews": "1,171",
+    "price": "\u20b91,760",
+    "link": "https://www.google.com/maps/place/GRAND+KALINGA+HOTEL/data=!4m10!3m9!1s0x3bae160e0ce07789:0xb15bf736f4238e6a!5m2!4m1!1i2!8m2!3d12.9762259!4d77.5786043!16s%2Fg%2F11sp32pz28!19sChIJiXfgDA4WrjsRao4j9Db3W7E?authuser=0&hl=en&rclk=1",
+    "amenities": [
+        "Pool available",
+        "Free breakfast available",
+        "Free Wi-Fi available",
+        "Free parking available"
+    ]
+}
+```
+
+### 5. Scroll to load more
+
+When scraping Google Maps, you'll notice that not all results load at once. Let's handle this infinite scroll pagination!
+
+First, we need a function that can handle the scrolling and detect when we've hit the bottom. Copy-paste this new function in the `gmap_scraper.py` file:
+
+```python
+async def load_more_items(page) -> bool:
```

Comment:
1. The article does not mention where the function should be called.
2. Crawlee already has `context.infinite_scroll()` - does it not work in this case?